### PR TITLE
Ability to combine multiple response projections #985

### PR DESCRIPTION
**Pull request by kobylynskyi**
---

### Description

Related to #985

---

Changes were made to:
- [X] Codegen library - Java
- [x] Codegen library - Kotlin
- [x] Codegen library - Scala

### CODE REVIEW

**Comment by kobylynskyi** [comment]
@jxnu-liguobin
I've made some changes to kotlin/scala `.ftl`s and would appreciate if you take a look at it and possibly make some adjustments too.
Thanks!

File: src/main/resources/templates/scala-lang/parametrized_input.ftl [hunk 1]
```diff
@@ -54,6 +54,16 @@ case class ${className}(
 </#if>
 ) extends GraphQLParametrizedInput {
 
+    override def deepCopy(): ${className} {
```
Comment (jxnu-liguobin) [inline src/main/resources/templates/scala-lang/parametrized_input.ftl:57]:
```suggestion
    override def deepCopy(): ${className} = {
```

File: src/main/resources/templates/scala-lang/parametrized_input.ftl [hunk 2]
```diff
@@ -54,6 +54,16 @@ case class ${className}(
...
+        val parametrizedInput = ${className}()
```
Comment (jxnu-liguobin) [inline src/main/resources/templates/scala-lang/parametrized_input.ftl:58]: Unable to use new, and then assign values ​​one by one.
should be
A(
field1,field2,....
)

File: src/main/resources/templates/kotlin-lang/parametrized_input.ftl [hunk 3]
```diff
@@ -38,6 +38,16 @@ data class ${className}(
 </#if>
 ) : GraphQLParametrizedInput {
 
+    override fun deepCopy(): ${className} {
+        val parametrizedInput = ${className}()
+    <#if fields?has_content>
```
Comment (jxnu-liguobin) [inline src/main/resources/templates/kotlin-lang/parametrized_input.ftl:43]: the same as scala.
should be
A(
field1,field2,....
)

File: src/main/resources/templates/scala-lang/response_projection.ftl [hunk 4]
```diff
@@ -27,7 +28,29 @@ import scala.collection.mutable.HashMap
 <#list annotations as annotation>
 @${annotation}
 </#list>
-class ${className} extends GraphQLResponseProjection {
+class ${className}() extends GraphQLResponseProjection() {
+
+    def this(projection: ${className}) = {
+        this()
+        if (projection != null) {
+            for (field <- projection.fields.values) {
```
Comment (jxnu-liguobin) [inline src/main/resources/templates/scala-lang/response_projection.ftl:36]:
```suggestion
            for (field <- projection.fields.values.asScala) {
```

File: src/main/resources/templates/scala-lang/response_projection.ftl [hunk 5]
```diff
@@ -27,7 +28,29 @@ import scala.collection.mutable.HashMap
...
+            }
+        }
+    }
+
+    def this(projections: scala.Seq[${className}]) = {
+        this()
+        if (projections != null) {
+            for (projection <- projections) {
+                if (projection != null) {
+                    for (field <- projection.fields.values) {
```
Comment (jxnu-liguobin) [inline src/main/resources/templates/scala-lang/response_projection.ftl:47]:
```suggestion
                    for (field <- projection.fields.values.asScala) {
```

File: src/main/resources/templates/scala-lang/response_projection.ftl [hunk 6]
```diff
@@ -78,13 +101,15 @@ class ${className} extends GraphQLResponseProjection {
...
     def ${field.methodName}(alias: String, input: ${field.parametrizedInputClassName} <#if field.type?has_content>, subProjection: ${field.type}</#if>): ${className} = {
-        fields.add(new GraphQLResponseField("${field.name}").alias(alias).parameters(input)<#if field.type?has_content>.projection(subProjection)</#if>)
+        add$(new GraphQLResponseField("${field.name}").alias(alias).parameters(input)<#if field.type?has_content>.projection(subProjection)</#if>)
         this
     }
 
 </#if>
 </#list>
 </#if>
+    override def deepCopy$(): ${className} = ${className}(this)
```
Comment (jxnu-liguobin) [inline src/main/resources/templates/scala-lang/response_projection.ftl:111]:
```suggestion
    override def deepCopy$(): ${className} = new ${className}(this)
```

**Review by jxnu-liguobin — CHANGES_REQUESTED** [review]
I left some comments, you can try to make changes, or wait for me to make changes sometime in the evening.

**Comment by jxnu-liguobin** [comment]
plz see https://github.com/kobylynskyi/graphql-java-codegen/pull/1032

**Comment by kobylynskyi** [comment]
@jxnu-liguobin thank you for your feedback, I've applied your changes to templates in this PR. Please take a look.

**Review by jxnu-liguobin — DISMISSED** [review]

**Pull request merged**